README.md
# NAME

DBIx::Class::InflateColumn::Geo - Inflate geometric columns to data structures

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10.1+-blue.svg" alt="Requires Perl 5.10.1+" />
    <a href="https://travis-ci.org/Csson/p5-DBIx-Class-InflateColumn-Geo"><img src="https://api.travis-ci.org/Csson/p5-DBIx-Class-InflateColumn-Geo.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/release/CSSON/DBIx-Class-InflateColumn-Geo-0.0100"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/DBIx-Class-InflateColumn-Geo/0.0100" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=DBIx-Class-InflateColumn-Geo%200.0100"><img src="http://badgedepot.code301.com/badge/cpantesters/DBIx-Class-InflateColumn-Geo/0.0100" alt="CPAN Testers result" /></a>
    </p>
</div>

# VERSION

Version 0.0100, released 2018-10-27.

# SYNOPSIS

    package TheSchema::Result::Park;
    use base 'DBIx::Class::Core';

    __PACKAGE__->load_components(qw/InflateColumn::Geo/);
    __PACKAGE__->add_columns({
        location => {
            data_type => 'point'
        },
        secondary_location => {
            data_type => 'point',
            geo_srid => 4326,
            geo_xname => 'longitude',
            geo_yname => 'latitude',
            geo_with_astext => 1,
        },
    });

Later:
    my $park = $schema->resultset('Park')->create({
        location => { x => 15.43, y => 54.32 },
        secondary\_location => { longitude => 12.32, latitude => 45.9843 },
    });
    say $park->location->{'x'};                  # 15.43
    say $park->secondary\_location->{'latitude'}; # 45.9843
    say $park->secondary\_location->{'astext'};   # POINT(12.32 45.9843)

    # Values can also be given as an array ref or as a space-separated string, both in x/longitude, y/latitude order:
    my $same_park = $schema->resultset('Park')->create({
        location => [15.43, 54.32],
        secondary_location => '12.32 45.9843',
    });

# DESCRIPTION

DBIx::Class::InflateColumn::Geo inflates geometry columns (so far, only `POINT` is supported) to more accessible data structures.

# COLUMN SPECIFICATION OPTIONS

Usage shown in the synopsis.

## geo\_srid

The spacial reference identifier you wish to use.

Default: `4326` (aka [WGS 84](http://spatialreference.org/ref/epsg/wgs-84/))

Set it to `undef` if you want to use your database's default.

## geo\_xname

The name you wish to use for the `X` (or `longitude`) value.

Default: `x`

## geo\_yname

The name you wish to use for the `Y` (or `latitude`) value.

Default: `y`

## geo\_with\_astext

A boolean determining whether the [Well-known text](https://en.wikipedia.org/wiki/Well-known_text) of the column is included in the inflated hash. It is not used during deflation.

Default: `0`

# COMPATIBILITY

I have only tested this on MariaDB 10.\*.

# SEE ALSO

- [DBIx::Class::GeomColumns](https://metacpan.org/pod/DBIx::Class::GeomColumns)

# SOURCE

[https://github.com/Csson/p5-DBIx-Class-InflateColumn-Geo](https://github.com/Csson/p5-DBIx-Class-InflateColumn-Geo)

# HOMEPAGE

[https://metacpan.org/release/DBIx-Class-InflateColumn-Geo](https://metacpan.org/release/DBIx-Class-InflateColumn-Geo)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for DBIx-Class-InflateColumn-Geo

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0100 2018-10-27T17:50:54Z UTC
 - Initial release

 [Stats & Dependency Changes]
 - (run req) + Carp (any)
 - (run req) + perl 5.010001
 - (test req) + ExtUtils::MakeMaker (any)
 - (test req) + File::Spec (any)
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (test req) + Test::More 0.96
 - (test rec) + CPAN::Meta 2.120900
 - (conf req) + ExtUtils::MakeMaker (any)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 6.012
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.008
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0200
 - (dev req) + Dist::Zilla::Plugin::CheckChangesHasContent 0.008
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::ConfirmRelease (any)
 - (dev req) + Dist::Zilla::Plugin::CopyFilesFromBuild 0.170880
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::ExecDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.045
 - (dev req) + Dist::Zilla::Plugin::Git::Check (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.014
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.034
 - (dev req) + Dist::Zilla::Plugin::Git::GatherDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Push (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Tag (any)
 - (dev req) + Dist::Zilla::Plugin::GithubMeta 0.54
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::License (any)
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::Manifest (any)
 - (dev req) + Dist::Zilla::Plugin::ManifestSkip (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::MetaJSON (any)
 - (dev req) + Dist::Zilla::Plugin::MetaNoIndex (any)
 - (dev req) + Dist::Zilla::Plugin::MetaProvides::Package 2.004003
 - (dev req) + Dist::Zilla::Plugin::MetaYAML (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PodSyntaxTests (any)
 - (dev req) + Dist::Zilla::Plugin::PodWeaver 4.006
 - (dev req) + Dist::Zilla::Plugin::Prereqs (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.055
 - (dev req) + Dist::Zilla::Plugin::Readme (any)
 - (dev req) + Dist::Zilla::Plugin::ReadmeAnyFromPod 0.163250
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::RunExtraTests 0.028
 - (dev req) + Dist::Zilla::Plugin::ShareDir (any)
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.057
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::ReportPrereqs 0.027
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Dist::Zilla::Plugin::TestRelease (any)
 - (dev req) + Dist::Zilla::Plugin::TravisYML 1.15
 - (dev req) + Dist::Zilla::Plugin::UploadToCPAN (any)
 - (dev req) + Pod::Elemental::Transformer::List (any)
 - (dev req) + Pod::Weaver::Plugin::SingleEncoding (any)
 - (dev req) + Pod::Weaver::Plugin::Transformer (any)
 - (dev req) + Pod::Weaver::PluginBundle::CorePrep (any)
 - (dev req) + Pod::Weaver::Section::Authors (any)
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::Collect (any)
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Pod::Weaver::Section::Generic (any)
 - (dev req) + Pod::Weaver::Section::Leftovers (any)
 - (dev req) + Pod::Weaver::Section::Legal (any)
 - (dev req) + Pod::Weaver::Section::Name (any)
 - (dev req) + Pod::Weaver::Section::Region (any)
 - (dev req) + Pod::Weaver::Section::Version (any)
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Pod 1.41
 - (dev req) + Test::Version 1
 - (dev req) + Test::Warnings 0.026
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0322


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Inflate geometric columns to data structures",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "DBIx-Class-InflateColumn-Geo",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "DBIx::Class::InflateColumn::Geo",
  "PREREQ_PM" => {
    "Carp" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0101",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/DBIx/Class/InflateColumn/Geo.pm
use 5.10.1;
use strict;
use warnings;

package DBIx::Class::InflateColumn::Geo;

# ABSTRACT: Inflate geometric columns to data structures
# AUTHORITY
our $VERSION = '0.0101';

use Carp qw/confess/;

sub register_column {
    my($self, $column, $info, @rest) = @_;

    $self->next::method($column, $info, @rest);

    return if !(defined $info->{'data_type'});

    $self->_handle_point($column, $info, @rest) if lc $info->{'data_type'} eq 'point';
}

sub _handle_point {
    my($self, $column, $info, @rest) = @_;

    my $srid = exists $info->{'geo_srid'} ? $info->{'geo_srid'} : 4326;
    my $xname = $info->{'geo_xname'} || 'x';
    my $yname = $info->{'geo_yname'} || 'y';
    my $with_astext = exists $info->{'geo_with_astext'} ? $info->{'geo_with_astext'} : 0;

    $self->inflate_column(
        $column => {
            inflate => sub {
                my($value, $object) = @_;

                my($astext) = $object->result_source->schema->storage->dbh->selectrow_array("SELECT ASTEXT(?)", {}, $value);

                if($astext =~ m{^POINT\(([^\s]+) ([^\s]+)\)$}i) {
                    my $result = {
                        $xname => $1,
                        $yname => $2,
                    };
                    if($with_astext) {
                        $result->{'astext'} = $astext;
                    }
                    return $result;
                }
                return;
            },
            deflate => sub {
                my($value, $object) = @_;

                my $stringified = ref $value eq 'HASH'  ? sprintf ('%s %s', $value->{ $xname }, $value->{ $yname })
                                : ref $value eq 'ARRAY' ? "$value->[0] $value->[1]"
                                :                         $value
                                ;

                if($stringified =~ m/[^-+\d\s\.]/) {
                    confess "The submitted value for column <$column> stringifies to <$stringified>, which is not suitable for POINT";
                }

                return $srid ? \"ST_PointFromText('POINT($stringified)', $srid)"
                             : \"ST_PointFromText('POINT($stringified)')"
                             ;
            },
        }
    );
}

1;

__END__

=pod

=encoding utf-8

=head1 SYNOPSIS

    package TheSchema::Result::Park;
    use base 'DBIx::Class::Core';

    __PACKAGE__->load_components(qw/InflateColumn::Geo/);
    __PACKAGE__->add_columns({
        location => {
            data_type => 'point'
        },
        secondary_location => {
            data_type => 'point',
            geo_srid => 4326,
            geo_xname => 'longitude',
            geo_yname => 'latitude',
            geo_with_astext => 1,
        },
    });

Later:
    my $park = $schema->resultset('Park')->create({
        location => { x => 15.43, y => 54.32 },
        secondary_location => { longitude => 12.32, latitude => 45.9843 },
    });
    say $park->location->{'x'};                  # 15.43
    say $park->secondary_location->{'latitude'}; # 45.9843
    say $park->secondary_location->{'astext'};   # POINT(12.32 45.9843)

    # Values can also be given as an array ref or as a space-separated string, both in x/longitude, y/latitude order:
    my $same_park = $schema->resultset('Park')->create({
        location => [15.43, 54.32],
        secondary_location => '12.32 45.9843',
    });

=head1 DESCRIPTION

DBIx::Class::InflateColumn::Geo inflates geometry columns (so far, only C<POINT> is supported) to more accessible data structures.

=head1 COLUMN SPECIFICATION OPTIONS

Usage shown in the synopsis.

=head2 geo_srid

The spacial reference identifier you wish to use.

Default: C<4326> (aka L<WGS 84|http://spatialreference.org/ref/epsg/wgs-84/>)

Set it to C<undef> if you want to use your database's default.

=head2 geo_xname

The name you wish to use for the C<X> (or C<longitude>) value.

Default: C<x>

=head2 geo_yname

The name you wish to use for the C<Y> (or C<latitude>) value.

Default: C<y>

=head2 geo_with_astext

A boolean determining whether the L<Well-known text|https://en.wikipedia.org/wiki/Well-known_text> of the column is included in the inflated hash. It is not used during deflation.

Default: C<0>

=head1 COMPATIBILITY

I have only tested this on MariaDB 10.*.

=head1 SEE ALSO

=for :list
* L<DBIx::Class::GeomColumns>

=cut
